osgp/protocol-adapter-iec60870/osgp-protocol-adapter-iec60870/src/main/java/org/opensmartgridplatform/adapter/protocol/iec60870/domain/services/ClientConnectionServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+        final ClientConnection cachedDeviceConnection = this.connectionCache.getConnection(deviceIdentification);
+
+        if (cachedDeviceConnection != null) {
+            LOGGER.info("Connection found in cache for device {}.", deviceIdentification);
+            return cachedDeviceConnection;
+        } else {
+            LOGGER.info("No connection found in cache for device {}, creating new connection.", deviceIdentification);
+            return this.createConnection(requestMetadata);
+        }
+    }
+
+    @Override
+    @PreDestroy
+    public void closeAllConnections() {
+        LOGGER.info("Closing all connections.");
+        final Collection<ClientConnection> connections = this.connectionCache.getConnections();
+        LOGGER.warn("{} active connections found, closing all.", connections.size());
+        connections.forEach(this::disconnect);
+    }
+
+    /**
+     * Closes the {@link Connection}, sends a disconnect request and closes the
+     * socket.
+     *
+     * @param deviceIdentification
+     *            Device for which to close the connection.
+     */
+    public void disconnect(final String deviceIdentification) {
+        final ClientConnection connection = this.connectionCache.getConnection(deviceIdentification);
+        this.disconnect(connection);
+    }
+
+    /**
+     * Closes the {@link Connection}, sends a disconnect request and closes the
+     * socket.
+     *
+     * @param deviceIdentification
+     *            Device for which to close the connection.
+     */
+    public synchronized void disconnect(final ClientConnection connection) {
+        final String deviceIdentification = connection.getConnectionParameters().getDeviceIdentification();
+        if (connection instanceof DeviceConnection) {
+
+            this.iec60870Client.disconnect(connection);
+
+            this.connectionCache.removeConnection(deviceIdentification);
+
+        } else {
+            LOGGER.warn("No connection found for deviceIdentification {}", deviceIdentification);
+        }
+    }
+
+    private ClientConnection createConnection(final RequestMetadata requestMetadata) throws ConnectionFailureException {
+        final ConnectionParameters connectionParameters = this.createConnectionParameters(requestMetadata);
+        final ResponseMetadata responseMetadata = ResponseMetadata.from(requestMetadata);
+
+        final ClientConnectionEventListener eventListener = new ClientConnectionEventListener(
+                connectionParameters.getDeviceIdentification(), this.connectionCache, this.asduHandlerRegistry,
+                responseMetadata);
+
+        final ClientConnection newDeviceConnection = this.iec60870Client.connect(connectionParameters, eventListener);
+
+        try {
+            this.connectionCache.addConnection(requestMetadata.getDeviceIdentification(), newDeviceConnection);
+        } catch (final ClientConnectionAlreadyInCacheException e) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1166)